### DISCUSSION TITLE
How can I add aliases?

### DISCUSSION THREAD

**Discussion by tars0x9752** [numtide/devshell#214] [Q&A] [answered]
Hi, I want to do something similar to the following in devshell with flake's devShells. I also use `nix-direnv`.

```nix
pkgs.mkShell {
  # ...

  shellHook = ''
    alias hi='echo hi'
  '';
}
```

At first I thought `devshell.startup.<name>.text` was equivalent of `shellHook`. So I tried something like the following, but the alias doesn't work.

```nix
pkgs.devshell.mkShell {
  devshell = {
    # ...

    startup.shell-hook.text = ''
      alias hi='echo hi'
    '';
  };

  commands = [
    # ...
  ];
}
```

Next I tried `devshell.interactive.<name>.text` too, but that didn't work either. I've just noticed It seems to work if I `nix develop` manually instead of `direnv`. But I want it work with `direnv`. Is that possible? or am I misunderstanding something? BTW, Thanks for the awesome library!

**Comment by zimbatm** [comment, marked as answer]
Because direnv targets many different types of shells and handles reverting the changes when switching projects, it can only handle environment variables. This unfortunately means that things like bash aliases and function definitions are being lost.

Most of the time aliases are being used as a shorthand to a longer program name. In that case, you can use the commands to generate a wrapper script:
```nix
pkgs.devshell.mkShell {
  commands = [
    { name = "tf"; category = "ops"; help = "terraform alias"; command = ''${tf}/bin/terraform "$@"''; }
  ];
}
```

**Reply by martinbaillie** [reply]
Anyone have any tricks for sourcing the command into the current shell like an alias would though?

e.g. `alias cdpr="cd $PRJ_ROOT"`

**Reply by tomdavidson** [reply]
If aliases are in `/etc/profile.d/`, can we just set `devshell.load_profiles=true` ?

What about aliases defined in .env/.envrc if direnv is used?

devshell.startup.alias.text = '. project-bashrc.sh' ?

**Comment by tars0x9752** [comment]
I see. Thanks for your answer!

**Comment by tomdavidson** [comment]
Bummer. The devshell command doesn't work for aliasing node with a loader and with tracing that you expect other programs invoking node to use.

**Comment by hugosenari** [comment]
For DSF I have alias 'alias' :rofl: module
Create a file with this and import in your project.nix
```nix
{pkgs, config, lib, ...}:
let
  cfg = config.alias;
  isntSH  = line: builtins.match "#!/.+" line == null;
  stripCo = line: builtins.replaceStrings ["# " "#"] ["" ""] line;
  lines   = value: lib.splitString "\n" value;
  toAlias = name: value: {
    name    = name;
    command = value;
    help    = with builtins; stripCo (head (filter isntSH (lines value)));
  };
  aliasses = builtins.attrValues (builtins.mapAttrs toAlias cfg);
in {
  options.alias = lib.mkOption {
    default       = {};
    description   = "bash script to create an alias";
    type          = lib.types.attrsOf lib.types.string;
    example.hello = "echo hello";
    example.world = ''
      #!/usr/bin/env python
      print("world")
    '';
  };
  config.commands = aliasses;
}
```

Then to set a new alias do this something like this:

```nix
{
   alias.hi="echo hi";
}
```

```toml
[alias]
hi = "echo hi"
```

The icing on the cake, create command with nushell to have a nice ux (--help, static typed args), builtin yaml, toml, json, etc
